src/idiomatic/leveraging-the-type-system/typestate-pattern.md

```diff
@@ -0,0 +1,89 @@
+---
+minutes: 30
+---
+
+## Typestate Pattern: Problem
+
+How can we ensure that only valid operations are allowed on a value based on its
+current state?
+
+```rust,editable
+use std::fmt::Write as _;
+
+#[derive(Default)]
+struct Serializer {
+    output: String,
+}
+
+impl Serializer {
+    fn serialize_struct_start(&mut self, name: &str) {
+        let _ = writeln!(&mut self.output, "{name} {{");
+    }
+
+    fn serialize_struct_field(&mut self, key: &str, value: &str) {
+        let _ = writeln!(&mut self.output, "  {key}={value};");
+    }
+
+    fn serialize_struct_end(&mut self) {
+        self.output.push_str("}\n");
+    }
+
+    fn finish(self) -> String {
+        self.output
+    }
+}
+
+fn main() {
+    let mut serializer = Serializer::default();
+    serializer.serialize_struct_start("User");
+    serializer.serialize_struct_field("id", "42");
+    serializer.serialize_struct_field("name", "Alice");
+
+    // serializer.serialize_struct_end(); // ← Oops! Forgotten
+
+    println!("{}", serializer.finish());
+}
+```
+
+<details>
+
+- This `Serializer` is meant to write a structured value. The expected usage
+  follows this sequence:
+
+```bob
+serialize struct start
+-+---------------------
+ |
+ +--> serialize struct field
+      -+---------------------
+       |
+       +--> serialize struct field
+            -+---------------------
+             |
+             +--> serialize struct end
+```
```

Comment:
I don't think we need this diagram. The intended flow here is pretty straightforward since the methods are intended to be used in the order they're declared. If you think it's worth explicitly calling out the intended flow then I think just writing it in a note would be enough.